duckduckgo/Android
# Add Anvil

Task/Issue URL: https://app.asana.com/0/414730916066338/1199916263225064/f
Tech Design URL: 
CC: 

**Description**:
We've been using Dagger Android for a while and it has been good help. We've also seen some drawbacks moving forward

* it is not easy to replace dependencies for testing, forcing us to create a `TestComponent`, custom `TestRunner` and a custom `TestApplication`
 * this also comes with some other problems like circular dependencies when splitting the codebase into multiple modules
* multi (gradle) module is harder to get right with just Dagger and Dagger-Android and we want to start splitting the app little by little

This PR:

- Adds Anvil and remove dagger-android annotation processor
- Creates a gradle module `:di` that is home to the Dagger components marker (scope) annotations
- Removes the TestApplication and TestComponent
- Adds a a bunch of binding classes under `com.duckduckgo.app.di.component`
 - these files replace the now removed `AndroidBindingModule.kt` that used the dagger-android annotation processor (DAAP)
 - DAAP can't work with Anvil because we need to replace the `@Component` and `@SubComponent` with `@MergeComponent` and `@MergeSubComponent` correspondingly and so we can't do that for the DAAP-generated code.
 - the codee in those files is pretty much the code generated by DAAP
  - the con is that we need to do it ourselves now, tho we can easily do this by either copy/paste/replace and/or a `file template` in AS (I did with a file template)
  - the pro is that it is now very clear what is under the hood, and so DI errors are less obscure now (hopefully)

Note:
* full build -> 1min40s
* incremental build -> 20s

**Steps to test this PR**:
1. App compiles
2. All tests pass
3. updgrade/install the app from this branch
4. verify that every entry that was in the `AndroidBindingModule.kt` has now a dedicated file in the package `com.duckduckgo.app.di.component`
  * eg. `BrowserActivity` -> `BrowserActivityComponent.kt`


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
